### PR TITLE
refactor(cli): move info subcommand over to new module graph

### PR DESCRIPTION
**Pull request by kitsonk**
This PR moves `deno info` over to the new module graph.

A couple of minor behavioural changes:

* The ordering of dependencies when displayed was not deterministic, as it relied upon the insertion in the module graph, which varied with the two.  Sort dependencies are not sorted.
* There is a line between the module information and the display of the tree. I can revert it, but I found it a lot clearer when viewing the display.
* The JSON serialisation now is ordered with the fields alphabetically, instead of the arbitrary order.
* Type only dependencies are reflected in the dependency tree (things such a `@deno-types`, `X-TypeScript-Types` and the triple slash types reference in JS files), as these are present in "natively" in the new module graph and are required to type check the code.

### CODE REVIEW

File: cli/module_graph2.rs [hunk 1]
```diff
@@ -421,6 +439,128 @@ impl Graph2 {
...
+    let module = self.modules.get(specifier).unwrap();
+    let mut deps = Vec::new();
+    let mut total_size = None;
+
+    if not_seen {
+      let mut seen_deps = HashSet::new();
+      // Check if the module has a type only dependency itself (e.g.
+      // `X-TypeScript-Types` header or triple slash ref)
+      if let Some((_, dep)) = &module.maybe_types {
+        if seen_deps.insert(dep.clone()) {
```
Comment (lucacasonato) [inline cli/module_graph2.rs:458]: I feel like type only dependencies should be separated in the JSON (type_deps field).

File: cli/module_graph2.rs [hunk 2]
```diff
@@ -421,6 +439,128 @@ impl Graph2 {
...
+      for (_, dep) in module.dependencies.iter() {
+        // Check the runtime code dependency
+        if let Some(code_dep) = &dep.maybe_code {
+          if seen_deps.insert(code_dep.clone()) {
+            deps.push(self.get_info(code_dep, seen, totals));
+          }
+        }
+        // Check the type only dependency (e.g. `@deno-types` pragma)
+        if let Some(type_dep) = &dep.maybe_type {
+          if seen_deps.insert(type_dep.clone()) {
```
Comment (lucacasonato) [inline cli/module_graph2.rs:471]: Ditto

File: cli/tests/info_recursive_imports_test.out [hunk 3]
```diff
@@ -1,11 +1,12 @@
 local: [WILDCARD]info_recursive_imports_test.ts
 type: TypeScript
 deps: 4 unique (total [WILDCARD])
+
 file://[WILDCARD]cli/tests/info_recursive_imports_test.ts ([WILDCARD])
 └─┬ file://[WILDCARD]cli/tests/recursive_imports/A.ts ([WILDCARD])
   ├─┬ file://[WILDCARD]cli/tests/recursive_imports/B.ts ([WILDCARD])
   │ ├─┬ file://[WILDCARD]cli/tests/recursive_imports/C.ts ([WILDCARD])
   │ │ ├── file://[WILDCARD]cli/tests/recursive_imports/A.ts *
-  │ │ └── file://[WILDCARD]cli/tests/recursive_imports/common.ts ([WILDCARD])
```
Comment (lucacasonato) [inline cli/tests/info_recursive_imports_test.out:9]: I am not fond of this change. Already displayed dependencies should be marked with *.

**Review by lucacasonato — COMMENTED** [review]
Very nice. Have some minor comments.

File: cli/tests/info_recursive_imports_test.out (hunk 3 above)
Comment (kitsonk) [reply cli/tests/info_recursive_imports_test.out:9]: It does, I was just finding that with the new module graph and circular dependencies, it wasn't deterministic which one got "seen" first.  It was a crap shoot.  I can try to figure out how to make it deterministic.

File: cli/module_graph2.rs (hunk 1 above)
Comment (kitsonk) [reply cli/module_graph2.rs:458]: If we are open to changing the JSON, I would like to rename quite a few fields actually, more aligned to the internals and consistent naming of things, as I think it really impedes maintainability for the long term.

File: cli/info.rs [hunk 4]
```diff
@@ -1,404 +1,232 @@
...
+use std::fmt;
+use std::path::PathBuf;
+
+/// Serialize a `MediaType` enum into a human readable string.  The default
+/// serialization for media types is and integer.
+fn serialize_media_type<S>(mt: &MediaType, s: S) -> Result<S::Ok, S::Error>
+where
+  S: Serializer,
+{
+  s.serialize_str(&format!("{}", mt))
```
Comment (ry) [inline cli/info.rs:22]: Presumably we want to do this everywhere we serialize MediaType. Maybe `impl Serialize for MediaType` where it's defined?

File: cli/info.rs [hunk 5]
```diff
@@ -1,404 +1,232 @@
...
+        compiled.to_string_lossy()
+      )?;
     }
     if let Some(ref map) = self.map {
-      f.write_fmt(format_args!("{} {}\n", colors::bold("map:"), map))?;
+      writeln!(f, "{} {}", colors::bold("map:"), map.to_string_lossy())?;
     }
-
-    f.write_fmt(format_args!(
-      "{} {} unique {}\n",
```
Comment (ry) [inline cli/info.rs:125]: It's odd to me that this module graph refactor is effecting the display. Are these changes unrelated edits?

File: cli/info.rs [hunk 6]
```diff
@@ -1,404 +1,232 @@
...
-  name: String,
-  size: usize,
-  total_size: Option<usize>,
-  deps: Vec<FileInfoDepTree>,
+pub struct ModuleInfo {
+  pub deps: Vec<ModuleInfo>,
+  #[serde(serialize_with = "serialize_specifier")]
+  pub name: ModuleSpecifier,
+  pub size: usize,
+  pub total_size: Option<usize>,
```
Comment (ry) [inline cli/info.rs:124]: I'd try to avoid superfluous renames/reordering of structs where possible to keep the diff small. I can't tell if this is related to the module graph or is just a better name/ordering.

File: cli/info.rs [hunk 7]
```diff
@@ -427,105 +252,89 @@ mod test {
...
-
-    let mut graph = ModuleGraph::new();
-
-    graph.insert(aspec.to_string(), a);
-    graph.insert(bspec.to_string(), b);
-    graph.insert(cspec.to_string(), c);
-
-    let count = get_unique_dep_count(&graph);
-
-    assert_eq!(graph.len() - 1, count);
```
Comment (ry) [inline cli/info.rs:529]: There are three tests here that got removed get_resolved_file_test, dependency_count_no_redirects, dependency_count_with_redirects. Why?

File: cli/info.rs (hunk 4 above)
Comment (kitsonk) [reply cli/info.rs:22]: No, because we have a serialize of `MediaType` that goes to the equivalent of the enum in JavaScript for `tsc`.  Though in the new infrastructure we don't serailize that at all to send into JavaScript.

I could refactor it so this is the default serialization and we modify the `module_graph` (legacy) to use a custom serializer, but I would just rather address it once we have completed the migration.

File: cli/info.rs (hunk 5 above)
Comment (kitsonk) [reply cli/info.rs:125]: I could choose to not refactor the display to more idiomatic Rust, but that seems non-sensical.  I really don't want to step around unmaintainable code when there is an opportunity to refactor it.

File: cli/info.rs (hunk 6 above)
Comment (kitsonk) [reply cli/info.rs:124]: Again, it seems totally non-sensical not to clean up stuff... otherwise it never gets cleaned up.

File: cli/info.rs (hunk 7 above)
Comment (kitsonk) [reply cli/info.rs:529]: I will take a look at those again, but they are testing behaviours that are not part of the new module graph, because of the way the graph is structure, but I will double check that the logic is covered.

File: cli/info.rs (hunk 5 above)
Comment (ry) [reply cli/info.rs:125]: Make its it easier to understand during complex refactors if only the necessary bits are touched - but no big deal

File: cli/info.rs [hunk 8]
```diff
@@ -1,404 +1,232 @@
...
+    writeln!(
+      f,
+      "{} {}",
+      self.info.name,
+      colors::gray(&format!("({})", human_size(self.info.size as f64)))
+    )?;
+
+    let dep_count = self.info.deps.len();
+    for (idx, dep) in self.info.deps.iter().enumerate() {
+      dep.write_info(f, &"", idx == dep_count - 1)?;
```
Comment (ry) [inline cli/info.rs:108]: replace `&""` with `""`

File: cli/info.rs [hunk 9]
```diff
@@ -1,404 +1,232 @@
...
 #[serde(rename_all = "camelCase")]
-struct FileInfoDepTree {
-  name: String,
-  size: usize,
-  total_size: Option<usize>,
-  deps: Vec<FileInfoDepTree>,
+pub struct ModuleInfo {
+  pub deps: Vec<ModuleInfo>,
+  #[serde(serialize_with = "serialize_specifier")]
+  pub name: ModuleSpecifier,
```
Comment (ry) [inline cli/info.rs:122]: https://github.com/denoland/deno/pull/7900

File: cli/module_graph2.rs (hunk 1 above)
Comment (kitsonk) [reply cli/module_graph2.rs:458]: I guess changing the behaviour is outside of the scope of this PR.  I opened #7927 which we can deal with post this PR.

File: cli/info.rs (hunk 7 above)
Comment (kitsonk) [reply cli/info.rs:529]: Ok, I have double checked.  `ModuleGraph2` handles re-directs in a significantly different way than the old module graph.  It depends upon the `SpecifierHandler` to mediate the redirects (which is the right layer for these to be abstracted in).  This means that `ModuleGraph2` is _unaware_ of what is going on in the background, which is perfectly fine.  I went through some older issues we had when moving to the "old" `ModuleGraph` having to deal with redirects and retested with `deno run --no-check` which uses the new `ModuleGraph2` and everything is handled properly.

There are quite a few unit tests for `ModuleGraph2`, having to deal with exotic sorts of dependencies, which is the right way to test such logic, and now `info.rs` is simply a "printout" of what is in the graph, versus having to interpret it.  That means these sorts of tests are no longer applicable.

File: cli/tests/info_recursive_imports_test.out (hunk 3 above)
Comment (kitsonk) [reply cli/tests/info_recursive_imports_test.out:9]: I tried, but it seems that the recursion and the new `ModuleGraph2` don't like to be deterministic.  One of the three versions of `common.ts` gets marked with its size, and the other 2 get marked with the ` *`, it just seems like the luck of the draw.

Let me know how much time you want me to try to figure it out, because the information is accurate, it just isn't deterministic.

**Review by ry — APPROVED** [review]
LGTM
